### PR TITLE
[Feature] DevStats Periodic Log Writer

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

- Mo `Systems Metrics` (approx 50-100) are currently written as `Part File` to S3 using `TAEWriter` or `CSVWriter`. 
- We want to introduce a new set of `Developer Stats/Metrics` (approx 500-1000) which will be written to the error log and eventually updated to `systems_table` for SQL querying. 
- Note: This PR only handles writing dev metrics to error logs. Parsing the error log and populating `systems_table` with Developer Stats entries is done using a separate pipeline mentioned here: https://github.com/matrixorigin/matrixone/issues/8196

## Final Log Output
```log
2023/03/15 02:53:52.337436 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 DN dd4dccb4-4d3c-41f8-b482-5251dc7a41bf SHARED stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 60, "hits": 60, "hit rate": 1, "mem reads": 60, "mem hits": 60, "mem hit rate": 1, "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.337584 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 CN dd1dccb4-4d3c-41f8-b482-5251dc7a41bf SHARED stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 200, "hits": 200, "hit rate": 1, "mem reads": 200, "mem hits": 200, "mem hit rate": 1, "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.337606 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 LOG 7c4dccb4-4d3c-41f8-b482-5251dc7a41bf ETL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.337733 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 DN dd4dccb4-4d3c-41f8-b482-5251dc7a41bf LOCAL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.337753 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 DN dd4dccb4-4d3c-41f8-b482-5251dc7a41bf ETL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.337766 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 CN dd1dccb4-4d3c-41f8-b482-5251dc7a41bf LOCAL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.338510 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 CN dd1dccb4-4d3c-41f8-b482-5251dc7a41bf ETL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.338544 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 LOG 7c4dccb4-4d3c-41f8-b482-5251dc7a41bf SHARED stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
2023/03/15 02:53:52.338569 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 LOG 7c4dccb4-4d3c-41f8-b482-5251dc7a41bf LOCAL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
```

## Instructions for using the `StatsExporter` for writing `Service Stats` to log periodically

1. Identify the Service `Stats` class to be exported (ie written to log). Ensure that, the Stats class internally uses `stats.Counter` instead of `atomic.Int64` for capturing stats.

```go
// MockService for demo
type MockService struct {
	stats *MockStats
}

type MockStats struct {
	reads stats.Counter
	hits  stats.Counter
}

func NewMockService() *MockService {
	return &MockService{
		stats: &MockStats{},
	}
}

func (d *MockService) Do() {
	d.stats.reads.Add(2)
	d.stats.hits.Add(1)
}

func (d *MockService) Stats() *MockStats {
	return d.stats
}
```

2. Create a new LogExporter class for the identified Stats class. This will serve as an adapter to export the Stats values in a loggable format.

```go
// LogExporter for the Mock Service declared above
type MockServiceLogExporter struct {
	service *MockService
}

func NewMockServiceLogExporter(service *MockService) LogExporter {
	return &MockServiceLogExporter{
		service: service,
	}
}

func (c *MockServiceLogExporter) Export() []zap.Field {
	var fields []zap.Field

	stats := c.service.Stats()

        // Returns current_window value and resets the current_window value to 0
	reads := stats.reads.SwapW(0)
	hits := stats.hits.SwapW(0)

	fields = append(fields, zap.Any("reads", reads))
	fields = append(fields, zap.Any("hits", hits))

	return fields
}
```

3. Initialize the `LogExporter` class near to `Stats/Service` class initialization.
```go
	// 1. Initialize service
	service := NewMockService()

	// 2. Initialize LogExporter for the service
	serviceLogExporter := NewMockServiceLogExporter(service)
```
 4. Register LogExporter to global stats registry using stats.Register()
```go
	// 3. Register LogExporter to the default stats registry
	stats.Register("MockServiceStats1", stats.WithLogExporter(serviceLogExporter))
```

Example Link: [registry_test.go](https://github.com/matrixorigin/matrixone/blob/b04e437dfc70f5093ed36a62657972f16610ff4f/pkg/util/metric/stats/registry_test.go)

### CODE REVIEW

File: cmd/mo-service/config.go [hunk 1]
```diff
@@ -184,6 +185,10 @@ func (c *Config) createFileService(defaultName string, perfCounter *perfcounter.
 		services = append(services, service)
 		perfCounterInfos = append(perfCounterInfos, counter)
 
+		// Create "Log Exporter" for this PerfCounter
```
Comment (arjunsk) [inline cmd/mo-service/config.go:188]: @reusee Right now, I am registering this global counter to the `statsRegistry` inside the `for loop`. It seems like this `counter.fsName` can be duplicated (eg, LOCAL). Could you suggest me a counter which has a unique familyName?

File: pkg/perfcounter/counter.go [hunk 2]
```diff
@@ -15,43 +15,42 @@
...
+		Read     stats.Counter
+		Hit      stats.Counter
+		MemRead  stats.Counter
+		MemHit   stats.Counter
+		DiskRead stats.Counter
+		DiskHit  stats.Counter
 	}
 
 	DistTAE struct {
 		MPool mpool.MPoolStats
```
Comment (fengttt) [inline pkg/perfcounter/counter.go:42]: I am still not quite sure if we should put mpoolstats here.    Not clean to me how do we use this ...

File: pkg/perfcounter/counter.go [hunk 3]
```diff
@@ -15,43 +15,42 @@
 package perfcounter
 
 import (
-	"sync/atomic"
-
 	"github.com/matrixorigin/matrixone/pkg/common/mpool"
+	"github.com/matrixorigin/matrixone/pkg/util/metric/stats"
 )
 
 type Counter struct {
```
Comment (fengttt) [inline pkg/perfcounter/counter.go:22]: Call it CounterSet?   Reduce name confusions.

File: pkg/perfcounter/counter.go (hunk 2 above)
Comment (arjunsk) [reply pkg/perfcounter/counter.go:42]: @reusee Please share your thoughts.

File: pkg/util/metric/stats/counter.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// Add adds to local counter
+func (c *Counter) Add(delta int64) (new int64) {
+	return c.local.Add(delta)
+}
+
+// Load return the sum of local and global
+func (c *Counter) Load() int64 {
+	return c.global.Load() + c.local.Load()
```
Comment (fengttt) [inline pkg/util/metric/stats/counter.go:32]: you can then load local and/or global

File: pkg/perfcounter/counter.go (hunk 3 above)
Comment (arjunsk) [reply pkg/perfcounter/counter.go:22]: I completely agree. @reusee please do the needful.

File: pkg/util/metric/stats/counter.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+// LoadC returns current local counter value.
+// TODO: Right now, only user of LoadC is LogExporter. Hence, MergeAndReset is also embedded here. It can be seperated later if required.
+func (c *Counter) LoadC() int64 {
```
Comment (fengttt) [inline pkg/util/metric/stats/counter.go:37]: and you can just swap.

LoadC is a bad name ...

File: pkg/util/metric/stats/counter.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+// Counter represents a combination of global & local counter.
+type Counter struct {
+	local  atomic.Int64
+	global atomic.Int64
+}
+
+// Add adds to local counter
+func (c *Counter) Add(delta int64) (new int64) {
+	return c.local.Add(delta)
```
[the remark this answers is not in the document]
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:27]: @fengttt  

That would be really good. By
1. Adding to only local, and Reading from local+global, we are pushing the extra cost to Reads.
2. Adding to local + global, and Reading from global alone, we are pushing the extra cost to Writes.

I found that writes were frequently called, and reads were only called for logging. Hence, I thought about having it this way.

File: pkg/util/metric/stats/counter.go (hunk 4 above)
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:32]: Right now, the default `Load()` exposes the global count only. In our case, `global count = global + local`.

`LoadC()` exposes the local counter value alone (it also merge and resets the value)

File: pkg/util/metric/stats/counter.go (hunk 5 above)
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:37]: Do you have suggestions for the name? Can I use `LoadLocal()` or `LoadL()` or `LoadWindow()` or `LoadW()`?

@fengttt 
I didn't understand "and you can just swap." Right now,
- this counter doesn't support Compare & Swap.
- LoadC() returns the current window value. It also does merging local counter value with the global one and then resets the local value.

Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:37]: Or sorry, I got it. You meant, call it Swap(), right? Making that change right away.

File: pkg/util/metric/stats/counter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:27]: But, the current approach violates the atomic nature. Do you want me to update this with "add to both local and global"?

[the remark this answers is not in the document]
Comment (fengttt) [reply pkg/util/metric/stats/counter.go:27]: yes, lets just add ...

File: pkg/util/metric/stats/counter.go (hunk 5 above)
Comment (fengttt) [reply pkg/util/metric/stats/counter.go]: I mean golang atomic can do a Swap

File: pkg/util/metric/stats/counter.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go:27]: Sure. I will add that.

File: pkg/util/metric/stats/counter.go (hunk 5 above)
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go]: Oh! Got it. Sure. Will make that change.

File: pkg/util/metric/mometric/metric.go [hunk 7]
```diff
@@ -80,6 +83,7 @@ func InitMetric(ctx context.Context, ieFactory func() ie.InternalExecutor, SV *c
 		moCollector = newMetricCollector(ieFactory, WithFlushInterval(initOpts.exportInterval))
 	}
 	moExporter = newMetricExporter(registry, moCollector, nodeUUID, role)
+	statsLogWriter = newStatsLogWriter(&stats.DefaultRegistry, runtime.ProcessLevelRuntime().Logger().Named("StatsLog"), metric.GetGatherInterval())
```
Comment (arjunsk) [inline pkg/util/metric/mometric/metric.go:87]: @xzxiong @aptend Can I use `runtime.ProcessLevelRuntime().Logger().Named("StatsLog")` here? I have updated the metric_test.go as well, with the relevant change.

File: pkg/util/metric/stats/stats_counter.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stats
+
+import "sync/atomic"
+
+// Counter represents a combination of global & local counter.
+type Counter struct {
```
[the remark this answers is not in the document]
Comment (arjunsk) [reply pkg/util/metric/stats/stats_counter.go]: Solved.

File: pkg/perfcounter/counter.go (hunk 3 above)
Comment (fengttt) [reply pkg/perfcounter/counter.go:22]: @arjunsk just change it in this PR>

File: pkg/util/metric/stats/counter.go (hunk 4 above)
Comment (arjunsk) [reply pkg/util/metric/stats/counter.go]: Right now, only exposing current window counter, using SwapW().

**Comment by arjunsk** [comment]
@aptend @xzxiong @fengttt @reusee @zhangxu19830126 Kindly review this PR and provide your feedback.

**Review by reusee — APPROVED** [review]

**Comment by arjunsk** [comment]
There seems to be a data race issue
image

Looking into it.

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/util/metric/mometric/metric.go (hunk 7 above)
Comment (xzxiong) [reply pkg/util/metric/mometric/metric.go:87]: sure, then the logger_name would be `cn-service.StatsLog` in CN, `dn-service.StatsLog` in DN, `log-service.StatsLog` in LOG.

File: pkg/util/metric/mometric/metric.go [hunk 9]
```diff
@@ -80,6 +83,7 @@ func InitMetric(ctx context.Context, ieFactory func() ie.InternalExecutor, SV *c
...
```
Comment (xzxiong) [inline pkg/util/metric/mometric/metric.go:86]: the`gatherInterval` args, I think it should have its own config

**Comment by xzxiong** [comment]
> ...
> 
> ## Final Log Output
> ```
> 2023/03/15 02:53:52.337436 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 DN dd4dccb4-4d3c-41f8-b482-5251dc7a41bf SHARED stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 60, "hits": 60, "hit rate": 1, "mem reads": 60, "mem hits": 60, "mem hit rate": 1, "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
> 2023/03/15 02:53:52.337584 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 CN dd1dccb4-4d3c-41f8-b482-5251dc7a41bf SHARED stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 200, "hits": 200, "hit rate": 1, "mem reads": 200, "mem hits": 200, "mem hit rate": 1, "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
> 2023/03/15 02:53:52.337606 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 LOG 7c4dccb4-4d3c-41f8-b482-5251dc7a41bf ETL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
> 2023/03/15 02:53:52.337733 -0500 INFO log-service.StatsLog mometric/stats_log_writer.go:86 DN dd4dccb4-4d3c-41f8-b482-5251dc7a41bf LOCAL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
> ```
> ...


@gavinyue  pls take a look at the value `NaN` of "hit rate".  Is that ok for stream engine to do fetch or calculate, maybe someday.

**Comment by arjunsk** [comment]
Fixing 
- Pointer Interface
- gatherInterval
- cn-service.StatsLog 
feedbacks.

File: pkg/util/metric/mometric/metric.go (hunk 7 above)
Comment (arjunsk) [reply pkg/util/metric/mometric/metric.go:87]: Hi @xzxiong , serviceName is not present inside metric.go. Did you mean, I should add serviceName prefix to `.Named("StatsLog")`? Is it ok to have `Named("StatsLog")` alone?

Comment (arjunsk) [reply pkg/util/metric/mometric/metric.go:87]: Thanks for confirming.

**Review by aptend — APPROVED** [review]

**Comment by gavinyue** [comment]
> e.StatsLog mometric/stats_log_writer.go:86 LOG 7c4dccb4-4d3c-41f8-b482-5251dc7a41bf ETL stats  {"uuid": "7c4dccb4-4d3c-41f8-b482-5251dc7a41bf", "reads": 0, "hits": 0, "hit rate": "NaN", "mem reads": 0, "mem hits": 0, "mem hit rate": "NaN", "disk reads": 0, "disk hits": 0, "disk hit rate": "NaN"}
> 2023/03/15 02:53:52.337733 -0500 INFO log-service.Stats

If you refer to whether later in streaming whether we can filter the NaN in the aggregate, the answer is yes.  We could filter out them.  

I am trying to understand why the hit is 0 and still outputs them. Is it because we want to diff from the system hanging and give the baseline that it is actually 0 hits?

**Comment by arjunsk** [comment]
> I am trying to understand why the hit is 0 and still outputs them. Is it because we want to diff from the system hanging and give the baseline that it is actually 0 hits?

The Dev Stats Logger requirement was to log, the counter/stats values every 15s.  Note that, the counter value logged is the value for that window and not the globalCounter values for the stats.

The globalCounter value can be calculated in the SQL using Sum().

I hope this helps.

**Pull request merged**